subgraph for hats-protocol

gershido opened the pull request:
First version. Creates the following entities:
Hat - includes all basic info of a hat, creation time, hat level, list of wearers, tree the hat belongs to
Wearer - includes account address, hats the account wears
Tree - includes list of hats that are in the tree

gershido committed:
delete files

gershido committed:
first version

spengrah commented on schema.graphql:
```diff
@@ -1,67 +1,25 @@
-  subHats: [HatTree!]
-}
-# Entity that recursively maps an Hat's tree
-type HatTree @entity {
```
Curious what you think about the utility of this entity. What would it enable for front ends? Do you think it makes sense to bring something like it back in a subsequent version?

spengrah commented on schema.graphql:
```diff
@@ -1,67 +1,25 @@
-  subHats: [HatTree!]
-}
-# Finite Hat data
 type Hat @entity {
```
Could also include the following attributes for the Hat entity, subject to cost-benefit analysis of including in the subgraph vs on the front end

- admin (retrievable from the id, but might be handy to have concretely, especially in combination with the removed HatTree entity -- eg to look up sibling hats)
- current supply (length of the wearers array)

spengrah replied:
What are your thoughts on also keeping track of changes to hat properties?

spengrah replied:
Oh, the other thing that would be amazing to have here would be a "pretty id", which formats an id kind of like an IP address...
- in hex, with the `0x` prefix
- periods between the hat levels
- only show the levels with non-zero values

For example, the pretty id of a level 3 hat might look like this: `00000001.01.01`

spengrah replied:
Maybe also a history of wearers of the hat?

spengrah commented on schema.graphql:
```diff
@@ -1,67 +1,25 @@
-  age: Int!
-  tx: TX!
-  allHats: [Hat!]
+  hats: [Hat!]! @derivedFrom(field: "wearers")
```
We'll also want to track

- badStandings (alternatively, create a new HatWearer entity with a standing attribute)
- a history of all hats worn by a given wearer

spengrah reviewed (commented):
Thanks so much for getting this started! My apologies for taking a while to review.

I left a few comments and questions in-line, plus a high level thought as well below. I won't merge quite yet, but can do so without further changes if the checkpoint is helpful for your process.

One challenge we'll have is that hat status ("is the hat on or off?") and hat wearer status ("are they currently wearing the hat?") can change without a tx. 
- is there any way to keep the subgraph up to date with these types of changes?
- if not, to avoid confusion we'll have to be explicit about the fact that the subgraph might be out of sync with the on-chain state
    - hat status on-chain state comes from `viewHat`
    - hat wearer status on-chain state comes from `isWearerOfHat` or `balanceOf`

gershido committed:
second version

gershido commented:
Hey! I have committed a second version with the following changes:
- Added the pretty id format
- The hat entity has now an "admin" and "subHats" fields. the "admin" links to the admin hat one level above, while the "subHats" links to the hats that are one level below. This allows to easily walk along a few levels in the tree (e.g. getting the siblings of a hat)  
- Added the "currentSupply" field to the hat entity
- Added event fields to track the history of a hat, tree or wearer. For example, to get the history of hats that a wearer had, you can get all the mint and burn events from the wearer entity. To get the history of a hat, you can get all the mint and burn events from the hat entity. 
- Added a README file with example queries

Regarding the "HatTree" entity that was included previously in the schema, I guess its purpose was to treat every hat as a root in its own sub tree and thus retrieving the hats in it more easily. While I didn't make a cost analysis, I still removed this entity because I think it complicates the subgraph unnecessarily. Because of how the hat IDs work, it's easy to get a sub-tree by filtering. I included an example of fetching a sub tree in the README.

Regarding the bad standings, the contract does not currently include events that signal the relevant changes. I guess that it relates to the challenge you mentioned regarding tracking the hat status and the hat wearer status. My understanding is that the toggle and eligibility contracts can be stateless. For example, if the eligibility for an address depends on that address having a balance of an ERC20 above X, then the eligibility contract might just call the "balanceOf" of the ERC20, check the condition and return the result without recording any state. In this case, I see two options to keep the subgraph synced with this virtual state:
1. Having a block handler that iterates over all the hats and wearers and checks their status
2. Having an off-chain service that checks this states and calls the "checkHatStatus" and "checkHatWearerStatus" when needed, paying the gas...

I think option 1 is practically infeasible in terms of cost, although I didn't check.
For eligibility and toggle contracts that do include state, we can wait to see which contracts will be used and get standardized, and then include them in the subgraph with templates.

spengrah commented on schema.graphql:
```diff
@@ -2,24 +2,158 @@ type Hat @entity {
   id: ID!
```
I love the pretty id! Though I think its probably best used as a reference (additional attribute) rather than the main id for a hat.

spengrah commented:
Looking really good! Really liking these changes, and thanks for adding a readme! I have a few thoughts in response to your notes.

> The hat entity has now an "admin" and "subHats" fields. the "admin" links to the admin hat one level above, while the "subHats" links to the hats that are one level below. This allows to easily walk along a few levels in the tree (e.g. getting the siblings of a hat)

This makes sense! I like this, and agree it obviates the need for the previous HatTree entity.

> Regarding the bad standings, the contract does not currently include events that signal the relevant changes. 

If I added an event to the contract, would this be easier to track in the subgraph? This should be a relatively simple event to add to the contract (if contract size contraints allow).

> My understanding is that the toggle and eligibility contracts can be stateless. For example, if the eligibility for an address depends on that address having a balance of an ERC20 above X, then the eligibility contract might just call the "balanceOf" of the ERC20, check the condition and return the result without recording any state. In this case, I see two options to keep the subgraph synced with this virtual state:
> 
> 1. Having a block handler that iterates over all the hats and wearers and checks their status
> 2. Having an off-chain service that checks this states and calls the "checkHatStatus" and "checkHatWearerStatus" when needed, paying the gas...
>  
> I think option 1 is practically infeasible in terms of cost, although I didn't check.
For eligibility and toggle contracts that do include state, we can wait to see which contracts will be used and get standardized, and then include them in the subgraph with templates.

This all sounds about right. My current thinking is that the subgraph can't be treated as a source of truth for wearer and hat status. Option 1 indeed sounds quite expensive, and even if we set up such an offchain service (option 2) it wouldn't be 100% reliable from a liveness perspective.

So my current sense is that the best thing to do is be explicit (via good schema labeling, in-line comments, and readme) that the subgraph is only tracking the most recently indexed state, not necessarily the current actual state. Front ends should directly query the wearer status and hat status from the contract.

What do you think?

gershido committed:
prettyId as additional field instead of primary ID

gershido commented:
I think you're right regarding the pretty ID and commited a change that adds an additional "prettyId" field while the primary ID is now the hat ID in hex format. Also agree regarding the hat status and wearer status. As long as the status depends on the state of some external contracts which the subgraph doesn't index, we can only show the state as recorded in the Hats contract right now. Adding an event for the changes in "badStandings" will enable to add this status in the wearer entity

spengrah reviewed (approved):
Looking great! merging...
